WordPress/gutenberg
# Blocks: Add blocks "slash" autocomplete on default block

Related: #834
Supersedes: #2284

This pull request seeks to enable a user to add a block by typing a slash `/` in a new default paragraph block, followed by a search parameter. Using arrow keys and click/enter on a block result will replace the paragraph block with a new block of the selected type.

![autocomplete](https://user-images.githubusercontent.com/1779930/29937133-93550e80-8e52-11e7-9a6f-9290e44f67dd.gif)

__Implementation notes:__

There's a bit of reinventing the wheel here, but unfortunately I wasn't able to reuse very much:

- I wanted to use [Downshift](https://github.com/paypal/downshift) to manage the typeahead behavior, but it does not support `contenteditable`, only `input`
- I wanted to use a utility for retrieving caret position, but [textarea-caret-position](https://github.com/component/textarea-caret-position) applies only to textarea and input, not `contenteditable`, and [Caret.js](https://github.com/ichord/Caret.js) is dependent on jQuery
- I wanted to use [DropdownMenu](https://github.com/WordPress/gutenberg/tree/master/components/dropdown-menu) for the arrow navigation of options, but I needed only menu behaviors, not the toggling button
- I wanted to use [Popover](https://github.com/WordPress/gutenberg/tree/master/components/popover) to manage positioning, but the position of the autocomplete menu cannot be determined relative a specific node (i.e. retrieving position of text caret requires custom logic)

There may still be opportunity for refactoring to consolidate some of these behaviors.

__Testing instructions:__

Verify that you can insert a new block from an empty paragraph block by typing slash, followed by a block search.

1. Navigate to Gutenberg > New Post
2. Click "Write your story"
3. Type `/`
4. Note that the autocomplete shows
5. Type a block search
6. Note that the autocomplete filters
7. Use arrow keys to navigate results
8. Click or press enter to select block
9. Note that the empty paragraph block is replaced with a new block of the selected type
